subsys/net/lib/coap/coap.c

```diff
@@ -133,48 +172,68 @@ static int parse_option(const struct coap_packet *cpkt,
 	delta = option_header_get_delta(opt);
 	len = option_header_get_len(opt);
 
-	/* In case 'delta' doesn't fit the option fixed header. */
-	r = decode_delta(delta, context, &delta);
-	if (r < 0) {
+	/* r == 0 means no more data to read from fragment, but delta
+	 * field shows that packet should contain more data, it must
+	 * be a malformed packet.
+	 */
+	if (r == 0 && delta > COAP_OPTION_NO_EXT) {
 		return -EINVAL;
 	}
 
-	opt_len += r;
+	if (delta > COAP_OPTION_NO_EXT) {
+		/* In case 'delta' doesn't fit the option fixed header. */
+		r = decode_delta(context, delta, &delta, &hdr_len);
+		if ((r < 0) || (r == 0 && len > COAP_OPTION_NO_EXT)) {
+			return -EINVAL;
+		}
 
-	/* In case 'len' doesn't fit the option fixed header. */
-	r = decode_delta(len, context, &len);
-	if (r < 0) {
-		return -EINVAL;
+		*opt_len += hdr_len;
 	}
 
-	opt_len += r + len;
+	if (len > COAP_OPTION_NO_EXT) {
+		/* In case 'len' doesn't fit the option fixed header. */
+		r = decode_delta(context, len, &len, &hdr_len);
```

Comment:
If we encountered a delta value of 13/14/15 in the previous call to decode_delta() then wouldn't hdr_len still have that value in it and be used erroneously if this call has a normal length value?

Reply:
shouldn't that case be returned as an error at line 187?

Reply:
I think that's why line 193 checks whether there's an extended option length before it calls decode_delta().

Reply:
@rbtchc yes. @mike-scott line 193 will verify that condition.